File: Algorithm/rotate_array.py
class Solution:
    def rotate(self,nums,k):
        def rev(num,begin,end):
            left = begin
            right = end
            while left < right:
                num[left],num[right-1] = num[right-1],num[left]
                left +=1
                right -=1
            else:
                return 
        length=len(nums)
        k = k%length
        rev(nums,0,length)
        rev(nums,0,k)
        rev(nums,k,length)
    def rotate_1(self,nums,k):
        print nums
        length = len(nums)
        k = k%length
        nums[:] = nums[-k:] + nums[:-k]
        print nums
        
  

        


if __name__ == '__main__':
    sol=Solution()
    nums=[1,2,3,4,5,6,7]
    k=3
    sol.rotate_1(nums,k)

File: Algorithm/CPP/rotate_array.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
using namespace std;



class Solution {
public:
	void rotate(vector<int>& nums, int k) {
		k = k % (nums.size());
		reverse(nums, 0, nums.size() - 1);
		reverse(nums, 0, k - 1);
		reverse(nums, k, nums.size() - 1);
	}
	void reverse(vector<int>& nums, int start, int end)
	{
		while (start < end)
		{
			int tmp = nums[start];
			nums[start] = nums[end];
			nums[end] = tmp;
			start++;
			end--;			
		}
	}
};
int  main()
{
	Solution sol;
	vector<int> nums = { 1, 2, 3, 4, 5, 6, 7};
	int k = 3;
	sol.rotate(nums, k);

	for (auto i : nums) cout << i << " ";
	cout << endl;
	system("pause");
	return 0;
}
